_posts/2018-07-10-Play Framework-in-Scala–Action-Composition.md

```diff
@@ -0,0 +1,256 @@
+```
+
+Don't pay attention to bodyParser, executationContext or the type parameter B.
+The main thing about this LoggingActionBuilder (and to any other ActionBuilder) is the method:
+
+```scala
+def invokeBlock[A](request: Request[A],
+                   block: Request[A] => Future[Result]): Future[Result] = {
+  Logger.debug(s"Just received the following request: $request")
+  block(request)
+}
+```
+
+This method is unimplemented on the ActionBuilder trait. It must be defined by us for each custom ActionBuilder we want to define.
+The method is important because it is where you describe the common functionality that the Actions build through it must have.
+In this case, the common functionality is:
+<small><strong>Line 3</strong></small>: Log the Request.
+<small><strong>Line 4</strong></small>: Do something else with the request.
+
+Where the "something else" is not yet defined on the LoggingActionBuilder. It is each Action stemming from this ActionBuilder, which, by providing a specific block function, can materialize <small><strong>Line 4</strong></small>.
+
+Consider the 2 following examples which are methods that return an Action and that could be inside a given controller. They would also be associated, on the conf/routes configuration file, with one HTTP method and URI pattern.
+
+```scala
+def action1(): Action[_] = LoggingActionBuilder { request =>
+  val userId: Option[String] = request.headers.get("userId")
+  userId.fold{
+    BadRequest("The request did not have a userId on the headers")
+  }{ userId =>
+    Ok(s"The user id is $userId")
+  }
+}
+```
+
+```scala
+def action2(): Action[_]  = LoggingActionBuilder { request =>
+  val itemToBuy = request.body.\("itemToBuy").validate[String]
+  itemToBuy.fold({
+    _ => BadRequest("No valid item to buy on the body of request.")
+  },{ item =>
+    Ok(s"You are buying $item")
+  })
+}
+```
+
+<strong>About them, we can say:</strong>
+1. They are Actions.
+2. They are different Actions, i.e., they do different things in response to a given request.
+3. They are built from the LoggingActionBuilder, via providing the block function which is referred to on the invokeBlock method of the LoggingActionBuilder.
+4. Most importantly, both action1() and action2(), when called, will have their request logged.
+
+<h4> Explanation of the syntax </h4>
+
+Creating an Action from an ActionBuilder consists on calling the apply method of the ActionBuilder. 
+
+```scala
+trait ActionBuilder[+R[_], B] {
+  final def apply(block: R[B] => Result): Action[B] = ...
+  ...
+}
+```
+
+This apply method takes as argument the block function which "materializes" the Action. On the examples above, the curl parenthesis are a call to this apply method, and the function inside is the specific block of the Action at hand.
+
+<h2> Changing the type parameter R[_] </h2>
+
+The left type parameter of an ActionBuilder constrains the signature of the block function in the invokeBlock method:
```

Comment:
I would say the first type parameter instead of left. Don't know why but I got confused with that at first glance.